Update JS to Dart (#4354)
dart-lang/site-www: src/_guides/language/coming-from/js-to-dart.md

as both languages share many concepts.

Like JavaScript, Dart runs on an event loop,
so both languages execute code in a similar way.
For example, asynchronous concepts like futures
(promises in JavaScript) and the `async/await` syntax are very similar.

Dart is strongly typed, unlike JavaScript.
If you have used with TypeScript or Flow,
this should simplify learning Dart.
If you've mostly worked with pure JavaScript,
it might be more of an adjustment.
With strong typing, Dart catches many errors before compiling
that might exist in JavaScript code.

Dart enables null safety by default.
JavaScript doesn't support null safety.
As a JavaScript developer,
it might take a while to learn how to write null safe code,
but the trade-off is better protection against
null reference exceptions that are detected even
before compiling Dart code. (Thereby avoiding those
dreaded `TypeError`s that occur when doing operations
on a JavaScript variable that turns out to be null.)

## Conventions and linting

JavaScript and Dart both have linting tools to enforce standard conventions.
While JavaScript offers many tools, standards, and configurations,

> Dart has one official layout conventions and includes a linter to make
> compliance effortless. Dart calls this linter the analyzer.

Reviewer: The analyzer is a lot more than just the linter, providing warnings and errors for misuse of language features,
code completion, actions like "Find usages", "Go to definition", "Fix X", etc.

```suggestion
Dart has official layout and styles conventions and includes
a linter to make compliance effortless. 
Support for this and other analysis functionality is provided by the analyzer.
```

To customize the lint rules for your project,
follow the [Customizing static analysis][] instructions.

Dart provides [`dart fix`][] to find and fix errors that the analyzer finds.

Dart also provides a code formatter similar to JS tools like [Prettier][].
To format code in any Dart project, run [`dart format`](/tools/dart-format) on
your command line. In Flutter, use `flutter format`.
The IDE plugins for Dart and Flutter also provide this ability.

Dart supports trailing commas for comma-separated lists of collections,
parameters, or arguments. When you add the trailing comma,
the formatter places each list item on its own line.
When you believe your list may have more items at a later date,
add the trailing comma. Avoid adding the trailing comma for the formatting
benefit alone.

> JavaScript supports trailing commas in list and map literals only.ß

Reviewer 2: Spurious "ß"

#### Numbers

Dart has three data types for holding numbers:

`num`
: The equivalent to the generic number type in JavaScript.

`int`
: Any numeric value without a fractional part.

`double`
: Any numeric value as a 64-bit (double-precision) floating point.

The Dart API includes all these types as classes.
Both the `int` and `double` types share `num` as their parent class:

<img
  src="/assets/img/guides/number-classes.png"
  alt="num subclasses Object and int and double each subclass num">

As Dart considers numbers as objects, numbers can expose their
own utility functions as object methods.
You don't need to use an additional object to apply a function to a a number.

For example, to round a `double` to an integer:

{:.include-lang}
```js
let rounded = Math.round(2.5);
```

{:.include-lang}
```dart
var rounded = 2.5.round();
```

#### Strings

Strings in Dart work like strings in JavaScript.
To write a string literal, enclose it in single (`'`) or double (`"`)
quotation marks.
The majority of Dart developers use single quotes,
but the language enforces no standard.
Use double quotation marks if you don't want to escape apostrophes or

> single quotes within the string.

Author: I just removed apostrophes. That's a confusing addition.

The equality and relational operators of both languages
work in the same way:

| Meaning                   | JavaScript operator | Dart operator    |
|---------------------------|---------------------|------------------|
| Strict equal              | `===`               | `indentical(,)`  |
| Abstract equal            | `==`                | `==`             |
| Strict not equal          | `!==`               | `!indentical(,)` |

> | Abstract not equal        | `!=`                | `!=`             |

Reviewer: It seems like Dart and JavaScript's `==` operators shouldn't be equated, as mentioned earlier Dart's `==` doesn't do automatic type conversions.

Also it doesn't seem it's a good idea to equate `===` to Dart's `identical`] as it seems most modern JavaScript code defaults to `===` but `identical` generally shouldn't be the default choice for determining equality.

assert((-value >>> 4) > 0); // Unsigned shift right
```

#### Conditional operator

Both Dart and JavaScript contain a conditional operator (`?:`)
for evaluating expressions.
Some developers refer to this operator as a ternary operator
as it takes three operands.

> Dart has another operator (`[]=`) that takes three operands.

Author: I think I have this updated correctly.

print(animal.age); // 5
```

To make the first cascade syntax null-aware, write it as `?..`.

{:.include-lang}
```dart
var result = maybePerson
    ?..employment = employer
    ..salary = salary;
```

Dart ignores the entire cascade if the `maybePerson` value is `null`.

## Collections

This section covers some collection types in Dart and compare them
to similar types in JavaScript.

### Lists

> Dart writes list literals as JavaScript arrays.

Reviewer: This text is a bit confusing to me. It seems to imply Dart uses JavaScript arrays somehow.

Perhaps:

```suggestion
Dart writes list literals similarly to JavaScript arrays.
```

{{site.alert.note}}
  The index operator (`[]`) returns a **nullable** value.
  The operator returns `null` if the map does not include

>   the lookup key.

Reviewer: Saying it returns a nullable value and it returns `null` seems confusing and a bit unnecessary. Can we just say one or the other, or connect them better somehow?

### Futures

`Future` is Dart's version of a `Promise`:

> an asynchronous operation that resolves at a later point.

Reviewer: This clarification makes sense and I think should still be addressed.

});
```

Similarly, futures can fail like promises.
Catching errors works the same as well:

{:.include-lang}
```js
httpResponseBody
  .then(...)
  .catch(err => {
    console.log(
      "Promise encountered an error
       before resolving."
    );
  });
```

{:.include-lang}
```dart
httpResponseBody
  .then(...)
  .catchError((err) {
    print(
      'Future encountered an error

>        before resolving.'

Reviewer: @atsansone These should still be fixed.

The `Stream` class has utility constructors for
creating streams from `Future`s or `Iterable`s,
or for creating streams that emit values at a timed interval.
To learn more, see the [`Stream`][] API page.

[`Stream`]: {{site.dart-api}}/dart-async/Stream-class.html

##### StreamController

> The utility class [StreamController][] can create and control streams.

Reviewer: Keep this in code syntax to be consistent with elsewhere:

```suggestion
The utility class [`StreamController`][] can create and control streams.
```

consumer.consume(); // Eating mice...
```

### Mixins

Mixins are used to share functionality between classes.
You can use the mixin's fields and methods in the class,
using their functionality as if it were part of the class.
A class can use multiple mixins. This helps when multiple classes share the
same functionality,
without needing to inherit from each other or share a common ancestor.

Use the `with` keyword to add one or more comma-separated mixins to a class.

JavaScript has no keyword equivalent. JavaScript can use `Object.assign`
to merge additional objects into an existing object, after instantiating.

The following examples show how JavaScript and Dart achieve similar behavior:

{:.include-lang}
```js
class Animal {}

// Defining the mixins
class Flyer {
  fly = () => console.log('Flaps wings');
}
class Walker {
  walk = () => console.log('Walks legs');
}
 
class Bat extends Animal {}
class Goose extends Animal {}
class Dog extends Animal {}

// Composing the class instances with
// their correct functionality.
const bat =
  Object.assign(
    new Bat(),

>     new Flyer()

Reviewer 2: This works, but it relies on the `fly`/`walk` "methods" being declared as fields instead.
If you wrote it as 
```js
class Flyer {
  fly() => console.log('Flaps wings');
}
```
instead, which would be the canonical way to write a JS class, the `assign` would not copy the method (because it's on the prototype and is not enumerable).
Might be a caveat worth mentioning.

Author: As this describes a specific caveat for JS, I'll leave it as is.

}
```

### Public and private members

Like JavaScript, Dart has no access modifier keywords:
all class members are public by default.

JavaScript will include private class members in the next

> practical revision of the EcmaScript standard.

Reviewer: This whole talk about the EcmaScript spec for private members not being complete is confusing and I'm not sure it should be included. Either way, it looks like PrivateIdentifiers were part of EcmaScript 2022 which has been released.

Reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

Author: I checked the reference specifically; privates are in stage 4, which means they will be in the next practical revision. That's why I phrased it as I did.

}
```

JavaScript uses the hash as a convention.
Dart's compiler enforces use of the underscore for this feature.

Dart makes private members private to the library, not the class.
This means that you can access private members from code in the same library.
By default, Dart limits access to private class members to code in the same file.
To expand the scope of a library beyond one file, add the `part` directive.
When possible, [avoid using `part`][]. Reserve using `part` for code generators.

[avoid using `part`]: /guides/libraries/create-library-packages#organizing-a-library-package

### Late variables

To indicate that Dart initializes class fields at a later point,
assign the `late` keyword to those class fields.
Those class fields remain non-nullable.

> Do this when a variable cannot be observed before being initialized.

Reviewer: This new sentence is confusing as every variable technically "cannot be observed before being initialized". I think the original text was trying to say something like "Do this when a variable does not need to be(or is not) observed/accessed right away and can be initialized later".

Author: Fair. Fixed.

}
```

In JavaScript, this is a convention,
but in Dart this is a full language feature
enforced by the compiler.

Dart makes private members private to the library, not the class.
This means that you can access private members from code in the same library.
By default, this works anywhere in the same file.
Private class members can be accessed from code in the same file.

> Using the `part` directive to expand the scope of a library scope a single file.

Reviewer 2: Word missing? Maybe "beyond a single files"?

> When possible, [avoid doing so using `part`][]. The common practice would be

Reviewer 2: Just ``[avoid using `part`]``. There are other reasons to use `part` files and you should avoid those too if you can.

while remaining non-nullable. This is useful
for cases where a variable is never observed
before being initialized, allowing it to be initialized later.
This differs from labeling the field as nullable:

* (Non-nullable) late fields cannot have null assigned at a later point.

* (Non-nullable) late fields throw a runtime error when
  accessed before they initialize. This should be avoided.

{:.include-lang}
```dart
class PetOwner {
  final String name;
  late final Pet _pet;
  PetOwner(this.name, String petName) {
    // Cyclic object graph, cannot set _pet before owner exists.
    _pet = Pet(petName, this);
  }
  Pet get pet => _pet;
}
class Pet {
  final String name;
  final PetOwner owner;
  Pet(this.name, this.owner);
}
```

Use `late` for local variables only if unclear code results
in the compiler being unable determine if the code initialized the variable.

```dart
doSomething(int n, bool capture) {
  late List<Foo> captures;
  if (capture) captures = [];
  for (var i = 0; i < n; i++) {
    var foo = something(i);
    if (capture) captures.add(foo);
  }
}

In the preceding example, the compiler does not know to assign

> `captures` if `capture` is true. In this case `capture` is true,

Reviewer 2: Drop "In this case `capture` is true,", we don't know that. The logic is sound whether it's true or not.
